modules/alia-async/src/qbits/alia/async.clj

```diff
@@ -99,23 +103,44 @@
            (Futures/addCallback
             rs-future
             (reify FutureCallback
-              (onSuccess [_ result]
-                (async/go
-                  (try
-                    (loop [rows (codec/result-set
-                                 (.get ^ResultSetFuture rs-future)
-                                 result-set-fn
-                                 row-generator
-                                 codec)]
-                      (when-let [row (first rows)]
-                        (when (async/>! ch row)
-                          (recur (rest rows)))))
-                    (catch Exception err
-                      (async/put! ch
-                                  (ex->ex-info err
-                                               {:query statement
-                                                :values values}))))
-                  (async/close! ch)))
+              (onSuccess [_ _rs]
+                (let [^ResultSet rs (.get rs-future)
+                      rows (.iterator rs)]
+                  (async/go
+                    (try
+                      (loop []
+                        (let [avail (.getAvailableWithoutFetching rs)
+                              ch-state
+                              (loop [avail avail]
+                                (if (pos? avail)
+                                  ;; we must make sure the parent chan
+                                  ;; is (still) open as the user might
+                                  ;; have interupted streaming closing
+                                  ;; the chan returned.
+                                  (if (async/>! ch
+                                                (codec/decode-row (.next rows)
+                                                                  row-generator
+                                                                  decode))
+                                    (recur (unchecked-dec-int avail))
+                                    ::closed)
+                                  ::open))]
+                          (when (and (= ch-state ::open)
+                                     (not (.isFullyFetched rs)))
+                            (when (zero? avail)
+                              (let [p (async/promise-chan)]
+                                (-> (.fetchMoreResults rs)
+                                    (Futures/addCallback
+                                     (reify FutureCallback
+                                       (onSuccess [_ _r] (async/put! p ::success))
+                                       (onFailure [_ _ex] (async/put! p ::error)))))
```

Comment:
Hey Max, what happens in the scenario of a failure for page-n here? Is that effectively muted, or is an ex-info passed back on the results chan?

It's been a while since I checked the internals (certainly not since row-generator introduced), but I specifically use execute-chan-buffered because of the possibility of any query that has n-results > fetch-size encountering a read-exception part way through a query.